components/formats-gpl/src/loci/formats/in/HarmonyReader.java

```diff
@@ -0,0 +1,783 @@
+    super.close(fileOnly);
+    if (!fileOnly) {
+      if (reader != null) {
+        reader.close();
+      }
+      reader = null;
+      planes = null;
+      metadataFiles.clear();
+    }
+  }
+
+  /**
+   * @see loci.formats.IFormatReader#openBytes(int, byte[], int, int, int, int)
+   */
+  @Override
+  public byte[] openBytes(int no, byte[] buf, int x, int y, int w, int h)
+    throws FormatException, IOException
+  {
+    FormatTools.checkPlaneParameters(this, no, buf.length, x, y, w, h);
+
+    if (getSeries() < planes.length && no < planes[getSeries()].length) {
+      Plane p = planes[getSeries()][no];
+
+      if (new Location(p.filename).exists()) {
+        if (reader == null) {
+          reader = new MinimalTiffReader();
+        }
+        reader.setId(p.filename);
+        reader.openBytes(0, buf, x, y, w, h);
+        reader.close();
+      }
+    }
+
+    return buf;
+  }
+
+  // -- Internal FormatReader API methods --
+
+  /* @see loci.formats.FormatReader#initFile(String) */
+  @Override
+  protected void initFile(String id) throws FormatException, IOException {
+    // make sure that we have the XML file and not a TIFF file
+
+    if (!checkSuffix(id, "xml")) {
+      Location parent = new Location(id).getAbsoluteFile().getParentFile();
+      Location xml = new Location(parent, XML_FILES[0]);
+      if (!xml.exists()) {
+        throw new FormatException("Could not find XML file " +
+          xml.getAbsolutePath());
+      }
+      initFile(xml.getAbsolutePath());
+      return;
+    }
+    else {
+      super.initFile(id);
+    }
+
+    // assemble list of other metadata/analysis results files
+
+    Location parent = new Location(currentId).getAbsoluteFile().getParentFile().getParentFile();
+    String[] list = parent.list();
+    for (String f : list) {
+      Location path = new Location(parent, f);
+      if (path.isDirectory()) {
+        String[] results = path.list();
+        for (String r : results) {
+          if (!f.equals("Images") || !checkSuffix(r, "tiff")) {
+            metadataFiles.add(new Location(path, r).getAbsolutePath());
+          }
+        }
+      }
+      else {
+        metadataFiles.add(path.getAbsolutePath());
+      }
+    }
+
+    // parse plate layout and image dimensions from the XML file
+
+    String xmlData = DataTools.readFile(id);
+    HarmonyHandler handler = new HarmonyHandler();
+    XMLTools.parseXML(xmlData, handler);
+
+    // sort the list of images by well and field indices
+
+    ArrayList<Plane> planeList = handler.getPlanes();
+
+    ArrayList<Integer> uniqueRows = new ArrayList<Integer>();
+    ArrayList<Integer> uniqueCols = new ArrayList<Integer>();
+    ArrayList<Integer> uniqueFields = new ArrayList<Integer>();
+    ArrayList<Integer> uniqueZs = new ArrayList<Integer>();
+    ArrayList<Integer> uniqueTs = new ArrayList<Integer>();
+    ArrayList<Integer> uniqueCs = new ArrayList<Integer>();
+
+    for (Plane p : planeList) {
```

Comment:
I'm not a Java expert but this looks like being O(n^2) which could be nasty if the list of planes is in the order of 1e5 which could be the case with large datasets. Wouldn't it be possible to use a HashSet to check if a row, col, etc already appeared in the plane list?